File: binary/c_parser/testparser.c
/**
* (C) 2020 Geotab Inc
* (C) 2018 Volvo Cars
*
* All files and artifacts in this repository are licensed under the
* provisions of the license provided by the LICENSE file in this repository.
*
* 
* Example of parser for a binary format VSS tree.
**/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdint.h>
#include <stdbool.h>
#include "cparserlib.h"

long currentNode;
long rootNode;
char* vspecfile;

char* getTypeName(nodeTypes_t type) {
    switch (type) { 
        case SENSOR:
            return "SENSOR";
        case ACTUATOR:
            return "ACTUATOR";
        case ATTRIBUTE:
            return "ATTRIBUTE";
        case BRANCH:
            return "BRANCH";
        default:
            printf("getTypeName: unknown type(%d)\n", type);
            return "unknown";
        break;
    } // switch
}

void showNodeData(long currentNode, int currentChild) {
        printf("\nNode: name = %s, type = %s, uuid = %s, validate = %d, children = %d,\ndescription = %s\n", VSSgetName(currentNode), getTypeName(VSSgetType(currentNode)), VSSgetUUID(currentNode), VSSgetValidation(currentNode), VSSgetNumOfChildren(currentNode), VSSgetDescr(currentNode));
        if (VSSgetNumOfChildren(currentNode) > 0)
            printf("Node child[%d]=%s\n", currentChild, VSSgetName(VSSgetChild(currentNode, currentChild)));
//        for (int i = 0 ; i < VSSgetNumOfAllowedElements(currentNode) ; i++)
//            printf("Allowed[%d]=%s\n", i, VSSgetAllowedElement(currentNode, i));
        printf("#allowed=%d\n", VSSgetNumOfAllowedElements(currentNode));
        char* dtype = VSSgetDatatype(currentNode);
        if (dtype != NULL)
            printf("Datatype = %s\n", dtype);
        char* tmp = VSSgetUnit(currentNode);
        if (tmp != NULL)
            printf("Unit = %s\n", tmp);
}

int main(int argc, char** argv) {

    vspecfile = argv[1];
    rootNode = VSSReadTree(vspecfile);

    char traverse[10];
    currentNode = rootNode;
    int currentChild = 0;
    showNodeData(currentNode, currentChild);
    printf("\nThe following parser commands are available: 'u'(p)p/'d'(own)/'l'(eft)/'r'(ight)/s(earch)/m(etadata subtree)/n(odelist)/(uu)i(dlist)/w(rite to file)/h(elp), or any other to quit\n");
    while (true) {
        printf("\n'u'/'d'/'l'/'r'/'s'/'m'/'n'/'i'/'w'/'h', or any other to quit: ");
        scanf("%s", traverse);
        switch (traverse[0]) {
            case 'u':  //up
                if (VSSgetParent(currentNode) != 0) {
                    currentNode = VSSgetParent(currentNode);
                    currentChild = 0;
                }
                showNodeData(currentNode, currentChild);
            break;
            case 'd':  //down
                if (VSSgetChild(currentNode, currentChild) != 0) {
                    currentNode = VSSgetChild(currentNode, currentChild);
                    currentChild = 0;
                }
                showNodeData(currentNode, currentChild);
            break;
            case 'l':  //left
                if (currentChild > 0) {
                    currentChild--;
                }
                showNodeData(currentNode, currentChild);
            break;
            case 'r':  //right
                if (currentChild < VSSgetNumOfChildren(currentNode)-1) {
                    currentChild++;
                }
                showNodeData(currentNode, currentChild);
            break;
            case 's':  //search for nodes matching path
            {
                char searchPath[MAXCHARSPATH];
                printf("\nPath to resource(s): ");
                scanf("%s", searchPath);
                searchData_t searchData[MAXFOUNDNODES];
                int foundResponses = VSSSearchNodes(searchPath, rootNode, MAXFOUNDNODES, searchData, true, true, 0, NULL, NULL);
                printf("\nNumber of elements found=%d\n", foundResponses);
                for (int i = 0 ; i < foundResponses ; i++) {
                    printf("Found node type=%s\n", getTypeName(VSSgetType((long)(&(searchData[i]))->foundNodeHandles)));
                    printf("Found node datatype=%s\n", VSSgetDatatype((long)(&(searchData[i]))->foundNodeHandles));
                    printf("Found path=%s\n", (char*)(&(searchData[i]))->responsePaths);
                }
            }
            break;
            case 'n':  //create node list file "nodelist.txt"
            {
                int numOfNodes = VSSGetLeafNodesList(rootNode, "nodelist.txt");
                printf("\nLeaf node list with %d nodes found in nodelist.txt\n", numOfNodes);
            }
            break;
            case 'i':  //create node list file "uuidlist.txt"
            {
                int numOfNodes = VSSGetUuidList(rootNode, "uuidlist.txt");
                printf("\nUUID list with %d nodes found in uuidlist.txt\n", numOfNodes);
            }
            break;
            case 'm':  //subtree metadata
            {
                char subTreePath[MAXCHARSPATH];
                printf("\nPath to subtree node: ");
                scanf("%s", subTreePath);
                int depth;
                printf("\nSubtree depth: ");
                scanf("%d", &depth);
                searchData_t searchData[MAXFOUNDNODES];
                int foundResponses = VSSSearchNodes(subTreePath, rootNode, MAXFOUNDNODES, searchData, false, false, 0, NULL, NULL);
                long subtreeNode = (long)(&(searchData[foundResponses-1]))->foundNodeHandles;
                char subTreeRootName[MAXCHARSPATH];
                strcpy(subTreeRootName, VSSgetName((long)(&(searchData[foundResponses-1]))->foundNodeHandles));
                for (int i = 1 ; i < depth ; i++) {
                    strcat(subTreeRootName, ".*");
                }
                foundResponses = VSSSearchNodes(subTreeRootName, subtreeNode, MAXFOUNDNODES, searchData, false, false, 0, NULL, NULL);
                printf("\nNumber of elements found=%d\n", foundResponses);
                for (int i = 0 ; i < foundResponses ; i++) {
                    printf("Node type=%s\n", getTypeName(VSSgetType((long)(&(searchData[i]))->foundNodeHandles)));
                    printf("Node path=%s\n", (char*)(&(searchData[i]))->responsePaths);
                    printf("Node validation=%d\n", VSSgetValidation((long)(&(searchData[i]))->foundNodeHandles));
                }
            }
            break;
            case 'h':  //help
                printf("\nTo traverse the tree, 'u'(p)p/'d'(own)/'l'(eft)/'r'(ight)/s(earch)/m(etadata subtree)/n(odelist)/(uu)i(dlist)/w(rite to file)/h(elp), or any other to quit\n");
            break;
            case 'w':  //write to file
                VSSWriteTree(vspecfile, rootNode);
            break;
            default:
                return 0;
        }  //switch
    } //while


    return 0;
} // main

File: binary/go_parser/testparser.go
/**
* (C) 2020 Geotab Inc
*
* All files and artifacts in this repository are licensed under the
* provisions of the license provided by the LICENSE file in this repository.
*
*
* Test parser for the binary VSS tree.
**/

package main

import (
    def "github.com/COVESA/vss-tools/binary/go_parser/datamodel"
    parser "github.com/COVESA/vss-tools/binary/go_parser/parserlib"
    "os"
//    "strings"
    "fmt"
)

var root *def.Node_t

func getTypeName(nodeType def.NodeTypes_t) string {
    switch (nodeType) { 
        case def.SENSOR:
            return "SENSOR"
        case def.ACTUATOR:
            return "ACTUATOR"
        case def.ATTRIBUTE:
            return "ATTRIBUTE"
        case def.BRANCH:
            return "BRANCH"
        default:
            fmt.Printf("getTypeName: unknown type(%d)\n", nodeType)
            return "unknown"
    } // switch
}

func showNodeData(currentNode *def.Node_t, currentChild int) {
        fmt.Printf("\nNode: name = %s, type = %s, uuid = %s, validate = %d, children = %d,\ndescription = %s\n", parser.VSSgetName(currentNode), getTypeName(parser.VSSgetType(currentNode)), parser.VSSgetUUID(currentNode), parser.VSSgetValidation(currentNode), parser.VSSgetNumOfChildren(currentNode), parser.VSSgetDescr(currentNode))
        if (parser.VSSgetNumOfChildren(currentNode) > 0) {
            fmt.Printf("Node child[%d]=%s\n", currentChild, parser.VSSgetName(parser.VSSgetChild(currentNode, currentChild)))
        }
        for i := 0 ; i < parser.VSSgetNumOfAllowedElements(currentNode) ; i++ {
            fmt.Printf("Allowed[%d]=%s\n", i, parser.VSSgetAllowedElement(currentNode, i))
        }
        dtype := parser.VSSgetDatatype(currentNode)
        if (dtype != 0) {
            fmt.Printf("Datatype = %d\n", dtype)
        }
        tmp := parser.VSSgetUnit(currentNode)
        if (len(tmp) != 0) {
            fmt.Printf("Unit = %s\n", tmp)
        }
}

func main() {
    if len(os.Args) != 2 {
        fmt.Printf("testparser command line: ./testparser filename\n")
	os.Exit(1)
    }
    root = parser.VSSReadTree(os.Args[1])
    fmt.Printf("VSS tree root name = %s\n", parser.VSSgetName(root))
    var traverse string
    fmt.Printf("\nTo traverse the tree, 'u'(p)p/'d'(own)/'l'(eft)/'r'(ight)/s(earch)/m(etadata subtree)/n(odelist)/(uu)i(dlist)/w(rite to file)/h(elp), or any other to quit\n")
    currentNode := root
    currentChild := 0
    for {
        fmt.Printf("\n'u'/'d'/'l'/'r'/'s'/'m'/'n'/'i'/'w'/'h', or any other to quit: ")
        fmt.Scanf("%s", &traverse)
        switch (traverse[0]) {
            case 'u':  //up
                if (parser.VSSgetParent(currentNode) != nil) {
                    currentNode = parser.VSSgetParent(currentNode)
                    currentChild = 0
                }
                showNodeData(currentNode, currentChild)
            case 'd':  //down
                if (parser.VSSgetChild(currentNode, currentChild) != nil) {
                    currentNode = parser.VSSgetChild(currentNode, currentChild)
                    currentChild = 0
                }
                showNodeData(currentNode, currentChild)
            case 'l':  //left
                if (currentChild > 0) {
                    currentChild--
                }
                showNodeData(currentNode, currentChild)
            case 'r':  //right
                if (currentChild < parser.VSSgetNumOfChildren(currentNode)-1) {
                    currentChild++
                }
                showNodeData(currentNode, currentChild)
            case 's':  //search for nodes matching path
            {
                var searchPath string
                fmt.Printf("\nPath to resource(s): ")
                fmt.Scanf("%s", &searchPath)
                searchData, foundResponses := parser.VSSsearchNodes(searchPath, root, parser.MAXFOUNDNODES, true, true, 0, nil, nil)
                fmt.Printf("\nNumber of elements found=%d\n", foundResponses)
                for i := 0 ; i < foundResponses ; i++ {
                    fmt.Printf("Found node type=%s\n", getTypeName(parser.VSSgetType(searchData[i].NodeHandle)))
                    fmt.Printf("Found node datatype=%d\n", parser.VSSgetDatatype(searchData[i].NodeHandle))
                    fmt.Printf("Found path=%s\n", searchData[i].NodePath)
                }
            }
            case 'n':  //create node list file "nodelist.txt"
            {
                numOfNodes := parser.VSSGetLeafNodesList(root, "nodelist.txt")
                fmt.Printf("\nLeaf node list with %d nodes found in nodelist.txt\n", numOfNodes)
            }
            break;
            case 'i':  //create node list file "uuidlist.txt"
            {
                numOfNodes := parser.VSSGetUuidList(root, "uuidlist.txt")
                fmt.Printf("\nUUID list with %d nodes found in uuidlist.txt\n", numOfNodes)
            }
            case 'm':  //subtree metadata
            {
                var subTreePath string
                fmt.Printf("\nPath to subtree node: ")
                fmt.Scanf("%s", &subTreePath)
                var depth int
                fmt.Printf("\nSubtree depth: ")
                fmt.Scanf("%d", &depth)
                searchData, foundResponses := parser.VSSsearchNodes(subTreePath, root, parser.MAXFOUNDNODES, false, false, 0, nil, nil)
                subtreeNode := searchData[foundResponses-1].NodeHandle
                subTreeRootName := parser.VSSgetName(searchData[foundResponses-1].NodeHandle)
                for i := 1 ; i < depth ; i++ {
                    subTreeRootName += ".*"
                }
                fmt.Printf("\nsubTreeRootName=%s\n", subTreeRootName)
                searchData, foundResponses = parser.VSSsearchNodes(subTreeRootName, subtreeNode, parser.MAXFOUNDNODES, false, false, 0, nil, nil)
                fmt.Printf("\nNumber of elements found=%d\n", foundResponses)
                for i := 0 ; i < foundResponses ; i++ {
                    fmt.Printf("Node type=%d\n", parser.VSSgetType(searchData[i].NodeHandle))
                    fmt.Printf("Node path=%s\n", searchData[i].NodePath)
                    fmt.Printf("Node validation=%d\n", parser.VSSgetValidation(searchData[i].NodeHandle))
                }
            }
            case 'h':  //help
                fmt.Printf("\nTo traverse the tree, 'u'(p)p/'d'(own)/'l'(eft)/'r'(ight)/s(earch)/m(etadata subtree)/n(odelist)/(uu)i(dlist)/w(rite to file)/h(elp), or any other to quit\n")
            case 'w':  //write to file
                parser.VSSWriteTree(os.Args[1], root)
            default:
                os.Exit(0)
        }  //switch
    } //for

}
